File: go/test/company_search_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/company-enrich-sdk"
	"github.com/voxgig-sdk/company-enrich-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestCompanySearchEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CompanySearch(nil)
		if ent == nil {
			t.Fatal("expected non-nil CompanySearchEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := company_searchBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "company_search." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set COMPANYENRICH_TEST_COMPANY_SEARCH_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		companySearchRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.company_search", setup.data)))
		var companySearchRef01Data map[string]any
		if len(companySearchRef01DataRaw) > 0 {
			companySearchRef01Data = core.ToMapAny(companySearchRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = companySearchRef01Data

		// LIST
		companySearchRef01Ent := client.CompanySearch(nil)
		companySearchRef01Match := map[string]any{}

		companySearchRef01ListResult, err := companySearchRef01Ent.List(companySearchRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, companySearchRef01ListOk := companySearchRef01ListResult.([]any)
		if !companySearchRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", companySearchRef01ListResult)
		}

	})
}

func company_searchBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "company_search", "CompanySearchTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read company_search test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse company_search test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"company_search01", "company_search02", "company_search03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("COMPANYENRICH_TEST_COMPANY_SEARCH_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"COMPANYENRICH_TEST_COMPANY_SEARCH_ENTID": idmap,
		"COMPANYENRICH_TEST_LIVE":      "FALSE",
		"COMPANYENRICH_TEST_EXPLAIN":   "FALSE",
		"COMPANYENRICH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["COMPANYENRICH_TEST_COMPANY_SEARCH_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["COMPANYENRICH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["COMPANYENRICH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCompanyEnrichSDK(core.ToMapAny(mergedOpts))
	}

	live := env["COMPANYENRICH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["COMPANYENRICH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/company_search/CompanySearchEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { CompanyEnrichSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('CompanySearchEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when COMPANYENRICH_TEST_LIVE=TRUE.
  afterEach(liveDelay('COMPANYENRICH_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = CompanyEnrichSDK.test()
    const ent = testsdk.CompanySearch()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.COMPANY_ENRICH_TEST_LIVE
    for (const op of ['list']) {
      if (maybeSkipControl(t, 'entityOp', 'company_search.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set COMPANY_ENRICH_TEST_COMPANY_SEARCH_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let company_search_ref01_data = Object.values(setup.data.existing.company_search)[0] as any

    // LIST
    const company_search_ref01_ent = client.CompanySearch()
    const company_search_ref01_match: any = {}

    const company_search_ref01_list = await company_search_ref01_ent.list(company_search_ref01_match)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/company_search/CompanySearchTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = CompanyEnrichSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['company_search01','company_search02','company_search03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['COMPANY_ENRICH_TEST_COMPANY_SEARCH_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'COMPANY_ENRICH_TEST_COMPANY_SEARCH_ENTID': idmap,
    'COMPANY_ENRICH_TEST_LIVE': 'FALSE',
    'COMPANY_ENRICH_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['COMPANY_ENRICH_TEST_COMPANY_SEARCH_ENTID']

  const live = 'TRUE' === env.COMPANY_ENRICH_TEST_LIVE

  if (live) {
    client = new CompanyEnrichSDK(merge([
      {
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.COMPANY_ENRICH_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
